Added custom Filter.

Blackskyliner opened the pull request:
Hi there, as I just wrote a TS3 Viewer I had the problem to use the cid inside the client to display the channel in an easy and smart way. As I didn't found another way to do it with current possibilities (a new filter was no choice for me as it is some kind of normalisation and I use GameQ in a very generalized way) I added the possibility to return a callable form the protocol, which defines Protocol specific normalisation which can not be created with the current system. 

Hope you'll like it and add it to your repo.

EDIT: Somehow the Exception fix #60 also got pushed... I am not so familiar with the Github interface...

Blackskyliner committed:
Fixed wrong parameter in Exception creation

Blackskyliner committed:
Added custom normalizer for Protocols

Austinb commented:
I actually built an overloadable property that you can add to the protocol class to tell normalize what to do on its own. 

If you check out GameQ_Protocols_Bf3 it has an example.  Basically you can overload the entire property or just part of it on __construct or something. I have not backported this to any older protocols.  It was something I added when bf3 first came out so the normalize filter could be paired down.

Please take a look.  I am ok with adding a callback method that each protocol can have for specific issues but as far as replacing keys with other keys it should be standard and the protocol class should just tell the normalize filter what to replace with what.

Thoughts?

Blackskyliner commented:
Yea, I know the normalize prop. But for example my ts3 addition will not translate the the single player but the array which holds the players thus having it keyed by their id's to easily access them e.g 

``` php
$player = $data['players'][0]
echo $data['teams'][$player['clid']]['gq_name']
```

This is not possible via property overloading but it is really useful for display or processing purpose. 

I will also push a ts3 protocol fix this day where the player.name property will be set right as the overloaded property has no mapping to clientnickname and the team has no mapping from channelname to name. 

I am mobile ATM so sorry for eventual formatting fails. 

Am 13.10.2012 um 00:55 schrieb Austin Bischoff notifications@github.com:

> I actually built an overloadable property that you can add to the protocol class to tell normalize what to do on its own.
> 
> If you check out GameQ_Protocols_Bf3 it has an example. Basically you can overload the entire property or just part of it on __construct or something. I have not backported this to any older protocols.  It was something I added when bf3 first came out so the normalize filter could be paired down.
> 
> Please take a look. I am ok with adding a callback method that each protocol can have for specific issues but as far as replacing keys with other keys it should be standard and the protocol class should just tell the normalize filter what to replace with what.
> 
> Thoughts?
> 
> —
> Reply to this email directly or view it on GitHub.

Blackskyliner committed:
Fixes teamspeak3 normalize settings

Blackskyliner committed:
Added team to players normaisation

Just like the teamspeak2 Protocol does.

Blackskyliner committed:
Fixed the by reference-failure.

The Callback now has to retunr the modified results which will be used
for the result data.

Blackskyliner committed:
Fixed normalize filters on teamspeak3 protocol

Blackskyliner commented:
So? accept, decline or discuss?

Austinb commented:
I like the idea of having an overloadable method that each protocol class can use to make specific changes like in ts3 for example. I think it should be some generic function name like normalize() and not specific to any one protocol class.  Also I think it might be better to move the overriding key-value pairs into that function as well to clean things up and separate the normalize from the rest of the protocol class.

So for example in the protocols_core class it would be public function normalize() {} and in the ts3 class it would overload that function.  For the invoking side it would need to be moved from core and put into the normalize filter class and that class would invoke any normalize() functions. The normalize filter class has access to the instance of the protocol so it can invoke it easily.  Note that normalize is still considered a filter and not part of the core functionality, although I do want to make it part of the core eventually and remove it as a filter option the project is not to that stage yet.

I can use your changes and make some changes I listed above and some other minor ones and push it back out or you can if you want.  Let me know if you want to do it or not.

Thoughts? Ideas?  I really like the idea it just needs a bit of refinement in my humble opinion.  Sorry for the lack of input, been swamped and only enough time for bug fixes as of late.

Austinb commented:
Also for this project I am used the upper-case versions of the constants for TRUE, FALSE & NULL.   So those will need to be changed back or I can do it before I will merge it into the v2 master.

Blackskyliner commented:
Yea I will do as you suggested, sounds even better than my interception method.
In the same breath I will also change the constants to be uppercase. (maybe this weekend ;))

Austinb commented:
Blackskyliner if you want some help with these please feel free to put this into a new branch and submit it over for a pull request so I can help.  I would like to implement the changes as the idea is a good one.  I know working on open-source projects can be the last on the list. 

Not sure you can reissue a pull for a new branch with this request or if you have to close this one and open a new pull request from the new branch.

Blackskyliner commented:
I will create a new one - branched. I had no time before new year and after the new year I were rarely at my mac. I close this and wil reissue a pull request (tomorrow) on an actual rebase and with the modification like discussed before. Therefore I will link the two issues through mentioning this issue in the new ticket.
